packages/gatsby/src/redux/reducers/html.ts

```diff
@@ -0,0 +1,123 @@
+import { ActionsUnion, IGatsbyState, IHtmlFileState } from "../types"
+
+const FLAG_DIRTY_CLEARED_CACHE = 0b0000001
+const FLAG_DIRTY_NEW_PAGE = 0b0000010
+const FLAG_DIRTY_PAGE_QUERY_CHANGED = 0b0000100 // TODO: this need to be PAGE_DATA and not PAGE_QUERY, but requires some shuffling
+const FLAG_DIRTY_BROWSER_COMPILATION_HASH = 0b0100000
+
+type PagePath = string
+
+function initialState(): IGatsbyState["html"] {
+  return {
+    trackedHtmlFiles: new Map<PagePath, IHtmlFileState>(),
+    browserCompilationHash: ``,
+  }
+}
+
+export function htmlReducer(
+  state: IGatsbyState["html"] = initialState(),
+  action: ActionsUnion
+): IGatsbyState["html"] {
+  switch (action.type) {
+    case `DELETE_CACHE`: {
+      if (action.cacheIsCorrupt) {
+        // `public` doesn't exist so we can start fresh
+        return initialState()
+      } else {
+        // we can't just clear the cache here - we want to keep track of pages, so we mark them all as "deleted"
+        // if they are recreated "isDeleted" flag will be removed
+        state.browserCompilationHash = ``
+        state.trackedHtmlFiles.forEach(htmlFile => {
+          htmlFile.isDeleted = true
+          // there was a change somewhere, so just in case we mark those files are dirty as well
+          htmlFile.dirty |= FLAG_DIRTY_CLEARED_CACHE
+        })
```

Comment:
What is the downside of just clearing the cache here? Is it only that we will re-build all HTML files? Or also that there could be some stale files left as a result?

I am not sure if it makes sense to have special handling for this case as it makes things more complicated in general. Ideally, files should be the derived artifact of our state but not the other way around?

Reply:
With code as-is - this is to remove .html / page-data files that are not recreated later on.

> Ideally, files should be the derived artifact of our state but not the other way around?

I'm not sure if this is the case of the other way around - I don't exactly use individual files and work backward to create state from them - instead this is
 
> We will blow away our state, but there are will be some files created from state we had before and we will need to keep part of that state to know what we need to clean up. If we want to start completely clean, we will need to do expensive `public` deletion (or traverse public dir and delete some files)

Not the hill I will die on, but we do have opportunity to use previous state to do targetted deletions instead of brute force deletions ;)